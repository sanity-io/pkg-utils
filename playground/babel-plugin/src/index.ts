/** @public */
export function defineThing(options: {name: string; age: number}) {
  const {name, ...restOptions} = options

  return {name, rest: restOptions}
}
