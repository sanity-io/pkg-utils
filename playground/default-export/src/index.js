function main() {
  return true
}

export default main

export const version = '1.0.0'
