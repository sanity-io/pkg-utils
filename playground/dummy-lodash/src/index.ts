import {uniqueId} from 'lodash'

/** @public */
export const id = uniqueId('dummy-lodash')
