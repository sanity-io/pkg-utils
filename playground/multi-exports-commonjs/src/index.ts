export * from './defineConfig'
export * from './definePlugin'
export * from './shared'
export * from './types'
