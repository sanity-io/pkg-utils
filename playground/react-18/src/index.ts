export {App} from './App'
export {Input} from './Input'
