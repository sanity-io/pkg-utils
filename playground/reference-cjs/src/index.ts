export const format = 'cjs'
export const runtime = 'any'
