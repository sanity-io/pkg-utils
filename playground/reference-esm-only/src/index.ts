/** @public */
export const format = process.env.PKG_FORMAT as string

/** @public */
export const runtime = process.env.PKG_RUNTIME as string

/** @public */
export const path = process.env.PKG_FILE_PATH as string
