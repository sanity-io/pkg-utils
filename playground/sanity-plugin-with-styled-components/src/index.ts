import {definePlugin, type Plugin} from 'sanity'
import {colorType} from './schema'

/** @public */
export const colorInput: Plugin = definePlugin({
  name: '@sanity/color-input',
  schema: {types: [colorType]},
})

export {ColorInput} from './LazyColorInput'
export type {ColorDefinition} from './schema'
