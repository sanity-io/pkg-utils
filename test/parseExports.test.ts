import {describe, expect, test} from 'vitest'

import {getPkgExtMap, PackageJSON, parseExports} from '../src/node'

describe('parseExports', () => {
  test('parse basic package.json', () => {
    const extMap = getPkgExtMap({legacyExports: false})

    const pkg: PackageJSON = {
      type: 'commonjs',
      name: 'test',
      version: '0.0.0-test',
      bin: {test: './dist/cli.js'},
      source: './src/index.ts',
      main: './dist/index.js',
      exports: {
        '.': {
          source: './src/index.ts',
          require: './dist/index.js',
          default: './dist/index.js',
        },
        './package.json': './package.json',
      },
    }

    const exports = parseExports({pkg, extMap, strict: true})

    expect(exports).toEqual([
      {
        _exported: true,
        _path: '.',
        types: undefined,
        source: './src/index.ts',
        browser: undefined,
        import: undefined,
        require: './dist/index.js',
        default: './dist/index.js',
      },
    ])
  })

  test('should throw if `package.json` is missing from the exports', () => {
    const extMap = getPkgExtMap({legacyExports: false})

    const pkg: PackageJSON = {
      type: 'commonjs',
      name: 'test',
      version: '0.0.0-test',
      bin: {test: './dist/cli.js'},
      source: './src/index.ts',
      main: './dist/index.js',
      exports: {
        '.': {
          source: './src/index.ts',
          require: './dist/index.js',
          default: './dist/index.js',
        },
      },
    }

    expect(() => parseExports({extMap, pkg, strict: true})).toThrow(
      '\n- package.json: `exports["./package.json"] must be declared.',
    )
  })

  test('parse package.json with browser files', () => {
    const extMap = getPkgExtMap({legacyExports: false})

    const pkg: PackageJSON = {
      type: 'module',
      name: 'test',
      version: '0.0.0-test',
      exports: {
        '.': {
          types: './dist/types/src/index.d.ts',
          browser: {
            source: './src/index.ts',
            require: './dist/index.cjs',
            import: './dist/index.js',
          },
          source: './src/index.ts',
          require: './dist/index.cjs',
          import: './dist/index.js',
          default: './dist/index.js',
        },
        './package.json': './package.json',
      },
    }

    const exports = parseExports({extMap, pkg, strict: true})

    expect(exports).toEqual([
      {
        _exported: true,
        _path: '.',
        types: './dist/types/src/index.d.ts',
        source: './src/index.ts',
        browser: {
          source: './src/index.ts',
          require: './dist/index.cjs',
          import: './dist/index.js',
        },
        import: './dist/index.js',
        require: './dist/index.cjs',
        default: './dist/index.js',
      },
    ])
  })

  test('package.json with multiple exports errors', () => {
    const extMap = getPkgExtMap({legacyExports: false})

    const pkg: PackageJSON = {
      type: 'commonjs',
      name: 'test',
      version: '0.0.0-test',
      source: './src/index.ts',
      exports: {
        '.': {
          types: './lib/src/index.d.ts',
          source: './src/index.ts',
          import: './lib/index.wrong.suffix',
          require: './lib/index.wrong.suffix',
          default: './lib/index.js',
        },
        './package.json': './not/package.json',
      },
      main: './lib/index.js',
      module: './lib/index.mjs',
      types: './lib/src/index.d.ts',
    }

    expect(() => parseExports({extMap, pkg, strict: true})).toThrow(
      '\n- package.json: mismatch between "main" and "exports.require". These must be equal.' +
        '\n- package.json: mismatch between "module" and "exports.import" These must be equal.' +
        '\n- package.json: `exports["./package.json"] must be "./package.json".' +
        '\n- package.json with `type: "commonjs"` - `exports["."].require` must end with ".js"' +
        '\n- package.json with `type: "commonjs"` - `exports["."].import` must end with ".mjs"',
    )
  })
})
