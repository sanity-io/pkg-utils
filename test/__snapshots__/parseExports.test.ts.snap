// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: false}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > .mjs file endings are mandatory when "type" is "commonjs" 1`] = `
[Error: 
- package.json: mismatch between "main" and "exports.require". These must be equal.
- package.json with \`type: "commonjs"\` - \`exports["."].import\` must end with ".mjs"]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > gracefully falls back if the browsers field is invalid 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/index.browser.mjs"
-       "require": "./lib/index.browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > it handles "browsers" if it only redirects "source" 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/index.browser.mjs"
-       "require": "./lib/index.browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > minimal CJS package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > minimal CJS-only package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" condition is required when legacyExports: true 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" must end with \`.esm.js\` 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'commonjs', legacyExports: true}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.js"
-       "require": "./lib/browser.cjs"
-     }
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.js"
-       "require": "./lib/browser.cjs"
-     }
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: false}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > gracefully falls back if the browsers field is invalid 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/index.browser.js"
-       "require": "./lib/index.browser.cjs"
-     }
-     "import": "./lib/index.js",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > it handles "browsers" if it only redirects "source" 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/index.browser.js"
-       "require": "./lib/index.browser.cjs"
-     }
-     "import": "./lib/index.js",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > minimal ESM package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > minimal ESM-only package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" condition is required when legacyExports: true 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" must end with \`.esm.js\` 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.js"
-       "require": "./lib/browser.cjs"
-     }
-     "import": "./lib/index.js",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.js"
-       "require": "./lib/browser.cjs"
-     }
-     "import": "./lib/index.js",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "import": "./lib/index.js",
-     "require": "./lib/index.cjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: 'module', legacyExports: true}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "import": "./lib/index.mjs",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: false}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > .mjs file endings are mandatory when "type" is "commonjs" 1`] = `
[Error: 
- package.json: mismatch between "main" and "exports.require". These must be equal.
- package.json with \`type: "commonjs"\` - \`exports["."].import\` must end with ".mjs"]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > gracefully falls back if the browsers field is invalid 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/index.browser.mjs"
-       "require": "./lib/index.browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > it handles "browsers" if it only redirects "source" 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/index.browser.mjs"
-       "require": "./lib/index.browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > minimal CJS package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > minimal CJS-only package.json requires a default condition when legacyExports: true 1`] = `
[Error: 
- package.json - \`exports["."].default\` should be set to "./dist/index.js" when "legacyExports" is true
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" condition is required when legacyExports: true 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > legacyExports: true > the top level "module" must end with \`.esm.js\` 1`] = `
[Error: 
- package.json: "module" should be "./dist/index.esm.js" when "legacyExports" is true]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "exports" key is required > maps "source" to "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/browser.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "main" field 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "exports" key is required > shows a best effort message if there is no "source" field either 1`] = `
[Error: 
- package.json: \`exports\` are missing, please set a minimal configuration, for example:
- "exports": {
-   ".": {
-     "source": "./src/index.js",
-     "default": "./dist/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "exports" key is required > uses the "browsers" field to specify a "browser" condition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "browser": {
-       "source": "./src/index.ts",
-       "import": "./lib/browser.mjs"
-       "require": "./lib/browser.js"
-     }
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "exports" key is required > uses the "main" and "source" fields to specify a suggested "exports" definition 1`] = `
[Error: 
- package.json: \`exports\` are missing, it should be:
- "exports": {
-   ".": {
-     "source": "./src/index.ts",
-     "import": "./lib/index.mjs",
-     "require": "./lib/index.js",
-     "default": "./lib/index.js"
-   },
-   "./package.json": "./package.json"
- }]
`;

exports[`parseExports({type: undefined, legacyExports: true}) > invalid packages > the "source" field should be removed when "exports" is set 1`] = `
[Error: 
- package.json: the "source" property can be removed, as it is equal to exports["."].source.]
`;
